<!doctype html>
<html>
	<!-- header definition -->
	<head>
		<title>ForCasT Browser</title>
		<!-- preload resources -->
    <link rel="preload" href="../../bin/bootstrap-4.4.1-dist/css/bootstrap.min.css" as="style">
    <link rel="preload" href="../../bin/bootstrap-4.4.1-dist/jquery-3.4.1.min.js" as="script">
		<script src="../../bin/bootstrap-4.4.1-dist/jquery-3.4.1.min.js"></script>
		<link rel="stylesheet" href="../../bin/bootstrap-4.4.1-dist/css/bootstrap.min.css">
		<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/select2/4.0.6-rc.0/css/select2.min.css" rel="stylesheet">
		<script src="https://cdnjs.cloudflare.com/ajax/libs/select2/4.0.6-rc.0/js/select2.min.js"></script>
		<link rel="stylesheet" href="css/landing-page.css">
		<script src="js/functions.js"></script>
		<link rel="icon" href="../../crispr-icon.png" sizes="100x100">
	</head>
  <!-- end header definition -->
	<body>
		<div class="content">
			<!-- pull in the navigation links -->
			<nav class="navbar navbar-expand-sm">
				<ul class="navbar-nav">
					<li class="nav-item active">
						<a class= "navigationLink" href="index.html">Home</a>
					</li>
					<li class="nav-item">
						<a class="selectedNavigation" href="about.html">About</a>
					</li>
					<li class="nav-item">
						<a class="navigationLink" href="tutorial.html">Tutorial</a>
					</li>
					<li class="nav-item">
						<a class="navigationLink" href="documentation.html">Documentation</a>
					</li>
				</ul>
			</nav>
			<!-- end navigation bar definition-->			
			<div class="container topPadding">
				<h1 style="color: white;">About ForCasT</h1>
				<p style="color: white;">CRISPR-mediated genome editing has enabled researchers to perform mutagenesis experiments with relative ease. Though tools exist for guide RNA (gRNA) design and predicting their off-target sites, there exists a need for a comprehensive platform to design, view, evaluate, store, and catalogue gRNAs and their associated primers. ForCasT (For Cas Tool) integrates existing open source tools such as JBrowse, Primer3, BLAST, bwa, and Silica to create a complete allele design and quality assurance pipeline. It is available both from a public github repository and as a Docker image, for ease of installation and portability. ForCasT stores data locally, eliminating the security and privacy concerns that make external web-based applications less desirable. For any organism of interest, ForCasT downloads the genome along with annotations for genes, transcripts, and regulatory regions from Ensembl and loads them as tracks into the genome browser, JBrowse. Users can then select specific regions of interest, specify PAM sequences and design gRNAs to target these regions.</p><p style="color: white;">Importantly, ForCasT provides the user with specificity scores for gRNAs and predicted off-target sites to assist gRNA choice and then saves the information for user-selected gRNAs along with custom notes to an internal Mongo database. Once the gRNAs have been added to the database, users can trigger Primer3 to design primers for both the wild-type and mutant alleles. Primers are checked for specificity in the genome using BLAST and an in silico PCR check is performed with Silica. Selected gRNAs and primers are displayed visually as JBrowse tracks and stored permanently in the database. This allows users to retrieve information related to an experiment and re-analyze the target, if necessary. Additionally, once a design has been finalized, ForCasT generates files containing the guides, primers, and exons for a gene, which can be edited by sequence annotation tools such as SnapGene or ApE (A plasmid Editor). If a design fails, unsuccessful guides can be marked and their scores, off-targets, and genomic context used to make an informed decision about redesign. Currently only deletion alleles are supported, but future versions will allow for the design of point mutations (variants), conditional alleles, and mutations in non-coding regions.</p>
			</div>
		</div>
		<div class="footer">
                        <p><a href="mailto:hillary.elrick@sickkids.ca">Contact</a></p>
                </div>
	</body>
</html>



